sn_networking/src/cmd.rs

```diff
@@ -983,24 +974,72 @@ impl SwarmDriver {
         let _ = self.quotes_history.insert(peer_id, quote);
     }
 
-    fn try_interval_replication(&mut self) -> Result<()> {
-        // get closest peers from buckets, sorted by increasing distance to us
-        let our_peer_id = self.self_peer_id.into();
-        let closest_k_peers = self
+    /// From all local peers, returns any within (and just exceeding) current get_range for a given key
+    pub(crate) fn get_filtered_peers_exceeding_range(
+        &mut self,
+        target_address: &NetworkAddress,
+    ) -> Vec<PeerId> {
+        let acceptable_distance_range = self.get_request_range();
+        let target_key = target_address.as_kbucket_key();
+
+        let peers = self
             .swarm
             .behaviour_mut()
             .kademlia
-            .get_closest_local_peers(&our_peer_id)
-            // Map KBucketKey<PeerId> to PeerId.
-            .map(|key| key.into_preimage());
-
-        // Only grab the closest nodes within the REPLICATE_RANGE
-        let mut replicate_targets = closest_k_peers
-            .into_iter()
-            // add some leeway to allow for divergent knowledge
-            .take(REPLICATION_PEERS_COUNT)
+            .get_closest_local_peers(&target_key)
+            .filter_map(|key| {
+                // here we compare _bucket_, not the exact distance.
+                // We want to include peers that are just outside the range
+                // Such that we can and will exceed the range in a search eventually
+                if acceptable_distance_range.ilog2() < target_key.distance(&key).ilog2() {
+                    return None;
+                }
+
+                // Map KBucketKey<PeerId> to PeerId.
+                let peer_id = key.into_preimage();
+                Some(peer_id)
+            })
             .collect::<Vec<_>>();
 
+        peers
+    }
+
+    /// From all local peers, returns any within current get_range for a given key
+    /// Excludes self
+    pub(crate) fn get_filtered_peers_exceeding_range_or_closest_nodes(
+        &mut self,
+        target_address: &NetworkAddress,
+    ) -> Vec<PeerId> {
+        let filtered_peers = self.get_filtered_peers_exceeding_range(target_address);
+        let closest_node_buffer_zone = CLOSE_GROUP_SIZE + close_group_majority();
+        if filtered_peers.len() >= closest_node_buffer_zone {
+            filtered_peers
```

Comment:
Aren't we limiting the range-based search to a number instead of a range here?
Wouldn't that just make the address mining attack possible again?

Reply:
with `get_filtered_peers_exceeding_range` we get all the local peers we have within the range, so we're returning that just now, unless there's not enough, then we just grab closest here.